File: src/agents/warp.ts
import { existsSync } from "node:fs";
import { getOrCreateObject, readJsonFile, writeJsonFile } from "../config/json.js";
import type { Detection, RunOpts, ToolId } from "../registry.js";
import { warn } from "../util/colors.js";
import { findBinaryIn } from "../util/detect.js";
import * as paths from "../util/paths.js";
import { hasOwner, removeOwner, writeOwner } from "../util/unified-block.js";

const CONTEXT_MODE_TOOLS = [
  "ctx_search",
  "ctx_execute",
  "ctx_execute_file",
  "ctx_batch_execute",
  "ctx_index",
  "ctx_fetch_and_index",
];
const CODEGRAPH_TOOLS = ["codegraph_explore"];

/** Detect if Warp / Oz Agent is installed. */
export function detect(): Detection {
  const hasCli = !!findBinaryIn("warp", paths.warpKnownBinDirs());
  const hasDesktop = paths.warpDesktopPaths().some((p) => existsSync(p));
  if (hasCli && hasDesktop) return { installed: true, source: "cli+desktop" };
  if (hasCli) return { installed: true, source: "cli" };
  if (hasDesktop) return { installed: true, source: "desktop" };
  if (process.env.NODE_ENV === "test" && existsSync(paths.warpPaths().dir)) {
    return { installed: true, source: "config" };
  }
  const dir = paths.warpPaths().dir;
  if (existsSync(dir) || existsSync(paths.warpPaths().mcpConfig)) {
    return { installed: true, source: "config" };
  }
  return { installed: false, source: "" };
}

/** Wire a tool into Warp / Oz Agent. */
export async function wire(tool: ToolId, opts: RunOpts): Promise<boolean> {
  switch (tool) {
    case "rtk":
      if (!opts.dryRun) installWarpRtkHook();
      return true;
    case "codegraph":
      if (!opts.dryRun) {
        configureWarpMcp("codegraph");
        writeOwner("warp", "codegraph");
      }
      return true;
    case "context-mode":
      if (!opts.dryRun) {
        configureWarpMcp("context-mode");
        writeOwner("warp", "context-mode");
      }
      return true;
    case "caveman":
      if (!opts.dryRun) writeOwner("warp", "caveman");
      return true;
    case "ponytail":
      if (!opts.dryRun) writeOwner("warp", "ponytail");
      return true;
    case "principles":
      if (!opts.dryRun) writeOwner("warp", "principles");
      return true;
    default:
      return false;
  }
}

/** Unwire a tool from Warp / Oz Agent. */
export async function unwire(tool: ToolId, _opts: RunOpts): Promise<boolean> {
  switch (tool) {
    case "rtk":
      removeWarpRtkHook();
      return true;
    case "codegraph":
      removeWarpMcp("codegraph");
      removeOwner("warp", "codegraph");
      return true;
    case "context-mode":
      removeWarpMcp("context-mode");
      removeOwner("warp", "context-mode");
      return true;
    case "caveman":
      removeOwner("warp", "caveman");
      return true;
    case "ponytail":
      removeOwner("warp", "ponytail");
      return true;
    case "principles":
      removeOwner("warp", "principles");
      return true;
    default:
      return false;
  }
}

/** Verify a tool is wired into Warp / Oz Agent. */
export function verify(tool: ToolId): boolean | null {
  switch (tool) {
    case "rtk":
      return hasWarpRtkHook();
    case "codegraph":
      return warpMcpHas("codegraph");
    case "context-mode":
      return warpMcpHas("context-mode");
    case "caveman":
      return hasOwner("warp", "caveman");
    case "ponytail":
      return hasOwner("warp", "ponytail");
    case "principles":
      return hasOwner("warp", "principles");
    default:
      return null;
  }
}

// ─── Hooks ───────────────────────────────────────────────────

type HookCfg = Record<string, unknown>;

function loadHooks(): { cfg: Record<string, unknown>; raw: string } | null {
  const p = paths.warpPaths();
  const raw = paths.readFile(p.hooksFile) ?? "";
  if (!raw) return { cfg: {}, raw: "" };
  try {
    const cfg = JSON.parse(raw) as Record<string, unknown>;
    if (cfg && typeof cfg === "object" && !Array.isArray(cfg)) return { cfg, raw };
  } catch {
    // Intentional: fall through, hooks.json exists but is unparseable -> return null so
    // callers never overwrite the file with a partially-rebuilt structure.
  }
  return null;
}

function saveHooks(cfg: Record<string, unknown>): void {
  const p = paths.warpPaths();
  writeJsonFile(p.hooksFile, cfg);
}

function addHookGroup(
  cfg: Record<string, unknown>,
  event: string,
  matcher: string,
  hookEntry: HookCfg,
): void {
  const arr = (cfg[event] as unknown[]) ?? [];
  const existing = arr as Record<string, unknown>[];
  const cmd = hookEntry.command as string;
  for (const g of existing) {
    if (g.matcher !== matcher) continue;
    const hooks = g.hooks as unknown[] | undefined;
    if (!Array.isArray(hooks)) continue;
    for (const h of hooks) {
      const hc = (h as Record<string, unknown>).command as string | undefined;
      if (hc === cmd) return;
    }
  }
  existing.push({ matcher, hooks: [hookEntry] });
  cfg[event] = existing as never;
}

function removeHookGroup(cfg: Record<string, unknown>, event: string, substr: string): void {
  const arr = cfg[event] as unknown[] | undefined;
  if (!Array.isArray(arr)) return;
  // Remove only the inner hook entries that match `substr`. Empty groups are dropped;
  // groups that still have user hooks are kept. The event key is deleted only when
  // no groups remain.
  const kept = arr.reduce<Record<string, unknown>[]>((acc, g) => {
    const hooks = (g as Record<string, unknown>).hooks as unknown[] | undefined;
    if (!Array.isArray(hooks)) {
      acc.push(g as Record<string, unknown>);
      return acc;
    }
    const keptHooks = hooks.filter((h) => {
      const c = (h as Record<string, unknown>).command as string | undefined;
      return !c?.includes(substr);
    });
    if (keptHooks.length > 0) {
      acc.push({ ...(g as Record<string, unknown>), hooks: keptHooks });
    }
    return acc;
  }, []);
  if (kept.length === 0) delete cfg[event];
  else cfg[event] = kept as never;
}

function hasHook(cfg: Record<string, unknown>, event: string, substr: string): boolean {
  const arr = cfg[event] as unknown[] | undefined;
  if (!Array.isArray(arr)) return false;
  for (const g of arr) {
    const hooks = (g as Record<string, unknown>).hooks as unknown[] | undefined;
    if (!Array.isArray(hooks)) continue;
    for (const h of hooks) {
      const c = (h as Record<string, unknown>).command as string | undefined;
      if (c?.includes(substr)) return true;
    }
  }
  return false;
}

function installWarpRtkHook(): void {
  const p = paths.warpPaths();
  paths.ensureDir(p.dir);
  const res = loadHooks();
  if (!res) {
    warn(
      `warp: skipping hook install; could not parse ${p.hooksFile}. Restore it or delete it and re-run 'toksave wire warp'.`,
    );
    return;
  }
  const tok = paths.toksaveAbs();
  const command = `${tok} rtk-hook warp`;
  addHookGroup(res.cfg, "PreToolUse", "Execute", { type: "command", command, timeout: 10 });
  saveHooks(res.cfg);
}

function removeWarpRtkHook(): void {
  const res = loadHooks();
  if (!res) return; // hooks.json unreadable; leave the file untouched.
  removeHookGroup(res.cfg, "PreToolUse", "rtk-hook");
  saveHooks(res.cfg);
}

function hasWarpRtkHook(): boolean {
  const res = loadHooks();
  if (!res) return false;
  return hasHook(res.cfg, "PreToolUse", "rtk-hook");
}

// ─── MCP ─────────────────────────────────────────────────────

function configureWarpMcp(toolId: string): void {
  const p = paths.warpPaths();
  paths.ensureDir(p.dir);
  const abs = paths.toksaveAbs();
  const cfg = (readJsonFile(p.mcpConfig) as Record<string, unknown>) ?? {};
  const servers = getOrCreateObject(cfg, "mcpServers");

  const args =
    toolId === "codegraph"
      ? ["runmcp", "--agent", "warp", "codegraph", "serve", "--mcp"]
      : ["runmcp", "--agent", "warp", toolId];
  const entry: Record<string, unknown> = {
    command: abs,
    args,
  };
  if (toolId === "codegraph") {
    entry.enabledTools = CODEGRAPH_TOOLS;
  } else if (toolId === "context-mode") {
    entry.enabledTools = CONTEXT_MODE_TOOLS;
  }

  (servers as Record<string, unknown>)[toolId] = entry;
  writeJsonFile(p.mcpConfig, cfg);
}

function removeWarpMcp(toolId: string): void {
  const p = paths.warpPaths();
  const cfg = readJsonFile(p.mcpConfig) as Record<string, unknown> | null;
  if (!cfg) return;
  const mcp = cfg.mcpServers as Record<string, unknown> | undefined;
  if (!mcp?.[toolId]) return;
  delete mcp[toolId];
  writeJsonFile(p.mcpConfig, cfg);
}

function warpMcpHas(toolId: string): boolean {
  const p = paths.warpPaths();
  const cfg = readJsonFile(p.mcpConfig) as Record<string, unknown> | null;
  if (!cfg) return false;
  const mcp = cfg.mcpServers as Record<string, unknown> | undefined;
  return !!mcp?.[toolId];
}

File: src/agents/warp.rs
use crate::agents::Agent;
use crate::registry::{Detection, RunOpts, ToolId};
use crate::util::detect::find_binary_in;
use crate::util::errors::{Result, ToksaveError};
use crate::util::json::{get_or_create_object, read_json_file, write_json_file, write_json_pruned};
use crate::util::paths::{
    toksave_abs, warp_cli_paths, warp_desktop_paths, warp_known_bin_dirs, warp_mcp_files,
    warp_paths,
};
use crate::util::unified_block::{has_owner, remove_owner, write_owner};
use serde_json::Value;
use std::path::{Path, PathBuf};

pub struct WarpAgent;

impl WarpAgent {
    pub fn new() -> Self {
        Self
    }
}

impl Default for WarpAgent {
    fn default() -> Self {
        Self::new()
    }
}

struct FileBackup {
    path: PathBuf,
    previous: Option<Vec<u8>>,
    existed: bool,
}

fn restore_backups(backups: &[FileBackup]) {
    for b in backups.iter().rev() {
        if b.existed {
            if let Some(bytes) = &b.previous {
                let _ = std::fs::write(&b.path, bytes);
            }
        } else {
            let _ = std::fs::remove_file(&b.path);
        }
    }
}

fn backup_file(path: &Path) -> FileBackup {
    let existed = path.exists();
    FileBackup {
        path: path.to_path_buf(),
        previous: if existed {
            std::fs::read(path).ok()
        } else {
            None
        },
        existed,
    }
}

fn mcp_file_has(path: &Path, tool: &str) -> bool {
    read_json_file(path)
        .ok()
        .flatten()
        .and_then(|c| c.get("mcpServers").cloned())
        .and_then(|m| m.get(tool).cloned())
        .is_some()
}

fn mcp_has_all(tool: &str) -> bool {
    let files = warp_mcp_files();
    !files.is_empty() && files.iter().all(|f| mcp_file_has(f, tool))
}

fn upsert_mcp_all(tool: &str, entry: Value) -> Result<()> {
    let files = warp_mcp_files();
    let mut backups = Vec::new();
    for path in files {
        backups.push(backup_file(&path));
        let result = (|| {
            let mut cfg = read_json_file(&path)?.unwrap_or_else(|| serde_json::json!({}));
            let servers = get_or_create_object(&mut cfg, "mcpServers");
            servers[tool] = entry.clone();
            write_json_file(&path, &cfg)
        })();
        if let Err(e) = result {
            restore_backups(&backups);
            return Err(e);
        }
    }
    Ok(())
}

fn remove_mcp_all(tool: &str) -> Result<()> {
    let files = warp_mcp_files();
    let mut backups = Vec::new();
    for path in files {
        if !path.exists() {
            continue;
        }
        backups.push(backup_file(&path));
        let result = (|| {
            if let Some(mut cfg) = read_json_file(&path)? {
                if let Some(mcp) = cfg.get_mut("mcpServers").and_then(|v| v.as_object_mut()) {
                    mcp.remove(tool);
                }
                write_json_pruned(&path, &cfg)?;
            }
            Ok::<(), ToksaveError>(())
        })();
        if let Err(e) = result {
            restore_backups(&backups);
            return Err(e);
        }
    }
    Ok(())
}

impl Agent for WarpAgent {
    fn detect(&self) -> Detection {
        let p = warp_paths();
        let cli = warp_cli_paths();
        let has_warp_bin = find_binary_in("warp", &warp_known_bin_dirs()).is_some();
        let has_oz_bin = find_binary_in("oz", &warp_known_bin_dirs()).is_some();
        let has_cli = has_warp_bin || has_oz_bin;
        let has_desktop = std::env::var("TOKSAVE_TEST").is_err()
            && warp_desktop_paths().iter().any(|p| p.exists());
        if has_cli && has_desktop {
            return Detection {
                installed: true,
                source: "cli+desktop".to_string(),
            };
        }
        if has_cli {
            return Detection {
                installed: true,
                source: "cli".to_string(),
            };
        }
        if has_desktop {
            return Detection {
                installed: true,
                source: "desktop".to_string(),
            };
        }
        let has_config =
            std::env::var("TOKSAVE_TEST").is_ok() && (p.dir.exists() || cli.dir.exists());
        let has_mcp = warp_mcp_files().iter().any(|f| f.exists());
        if has_config || has_mcp {
            Detection {
                installed: true,
                source: "config".to_string(),
            }
        } else {
            Detection {
                installed: false,
                source: String::new(),
            }
        }
    }

    fn wire(&self, tool: ToolId, opts: &RunOpts) -> Result<bool> {
        if opts.dry_run {
            return Ok(true);
        }
        let p = warp_paths();

        match tool {
            ToolId::Codegraph => {
                upsert_mcp_all(
                    "codegraph",
                    serde_json::json!({
                        "command": toksave_abs(),
                        "args": ["runmcp", "--agent", "warp", "codegraph", "serve", "--mcp"]
                    }),
                )?;
                write_owner("warp", "codegraph")?;
                Ok(true)
            }
            ToolId::ContextMode => {
                upsert_mcp_all(
                    "context-mode",
                    serde_json::json!({
                        "command": toksave_abs(),
                        "args": ["runmcp", "--agent", "warp", "context-mode"]
                    }),
                )?;
                write_owner("warp", "context-mode")?;
                Ok(true)
            }
            ToolId::Caveman => {
                write_owner("warp", "caveman")?;
                Ok(true)
            }
            ToolId::Rtk => {
                // Trust boundary: if hooks_file exists but is corrupted, read_json_file will return ToksaveError::Config
                if p.hooks_file.exists() {
                    let raw = std::fs::read_to_string(&p.hooks_file).unwrap_or_default();
                    if !raw.trim().is_empty()
                        && serde_json::from_str::<serde_json::Value>(&raw).is_err()
                    {
                        return Err(ToksaveError::config(
                            &p.hooks_file.to_string_lossy(),
                            "Corrupted JSON in Warp hooks file",
                        ));
                    }
                }
                let mut cfg =
                    read_json_file(&p.hooks_file)?.unwrap_or_else(|| serde_json::json!({}));
                let hook_entry = serde_json::json!({
                    "matcher": "Execute",
                    "hooks": [{ "type": "command", "command": format!("{} rtk-hook warp", toksave_abs()), "timeout": 10 }]
                });
                crate::util::json::merge_pretool_use(&mut cfg, hook_entry, "rtk-hook warp");
                write_json_file(&p.hooks_file, &cfg)?;
                Ok(true)
            }
            ToolId::Ponytail => {
                write_owner("warp", "ponytail")?;
                Ok(true)
            }
            ToolId::Principles => {
                write_owner("warp", "principles")?;
                Ok(true)
            }
        }
    }

    fn unwire(&self, tool: ToolId, _opts: &RunOpts) -> Result<bool> {
        let p = warp_paths();
        match tool {
            ToolId::Codegraph => {
                remove_mcp_all("codegraph")?;
                remove_owner("warp", "codegraph")?;
                Ok(true)
            }
            ToolId::ContextMode => {
                remove_mcp_all("context-mode")?;
                remove_owner("warp", "context-mode")?;
                Ok(true)
            }
            ToolId::Caveman => {
                remove_owner("warp", "caveman")?;
                Ok(true)
            }
            ToolId::Rtk => {
                if let Some(mut cfg) = read_json_file(&p.hooks_file)? {
                    crate::util::json::remove_pretool_use(&mut cfg, "rtk-hook warp");
                    write_json_pruned(&p.hooks_file, &cfg)?;
                }
                Ok(true)
            }
            ToolId::Ponytail => {
                remove_owner("warp", "ponytail")?;
                Ok(true)
            }
            ToolId::Principles => {
                remove_owner("warp", "principles")?;
                Ok(true)
            }
        }
    }

    fn verify(&self, tool: ToolId) -> Option<bool> {
        let p = warp_paths();
        match tool {
            ToolId::Codegraph => Some(mcp_has_all("codegraph")),
            ToolId::ContextMode => Some(mcp_has_all("context-mode")),
            ToolId::Caveman => Some(has_owner("warp", "caveman")),
            ToolId::Rtk => {
                let hcfg = read_json_file(&p.hooks_file).ok().flatten();
                Some(hcfg.as_ref().is_some_and(|c| {
                    crate::util::json::has_pretool_with_command_marker(c, "rtk-hook warp")
                }))
            }
            ToolId::Ponytail => Some(has_owner("warp", "ponytail")),
            ToolId::Principles => Some(has_owner("warp", "principles")),
        }
    }
}
